src/frontend/src/flows/welcome.ts

```diff
@@ -0,0 +1,105 @@
+import { render, html, TemplateResult } from "lit-html";
+import { registerIfAllowed } from "../utils/registerAllowedCheck";
+import { setUserNumber } from "../utils/userNumber";
+import { unreachable } from "../utils/utils";
+import { Connection } from "../utils/iiConnection";
+import { LoginFlowResult, LoginData } from "../flows/login/flowResult";
+import { displayError } from "../components/displayError";
+import { footer } from "../components/footer";
+import { icLogo } from "../components/icons";
+import { auth } from "../flows/login";
+
+/** Properties of the "Welcome" screen, which welcomes new users. */
+type WelcomeProps = {
+  register: () => void;
+  signin: () => void;
+};
+
+const learnMore = html`
+  <p class="t-paragraph t-centered l-stack">
+    <a
+      class="t-link"
+      href="https://medium.com/dfinity/internet-identity-the-end-of-usernames-and-passwords-ff45e4861bf7"
+      target="_blank"
+      rel="noopener noreferrer"
+      >Learn More</a
+    >
+  </p>
+`;
+
+const pageContent = ({
+  register,
+  signin,
+}: WelcomeProps): TemplateResult => html` <section
+    class="l-container c-card c-card--highlight"
+    aria-label="Authentication"
+  >
+    <div class="c-logo">${icLogo}</div>
+    <p class="t-paragraph t-centered">Unlock the full power of Web3</p>
+    <article class="l-stack">
+      <div class="l-stack">
+        <button
+          type="button"
+          @click=${register}
+          id="registerButton"
+          class="c-button"
+        >
+          Create an Anchor
+        </button>
+        <button
+          type="button"
+          @click=${signin}
+          id="loginButton"
+          class="c-button c-button--secondary"
+        >
+          Use Existing
+        </button>
+      </div>
+    </article>
+
+    ${learnMore}
+  </section>
+  ${footer}`;
+
+export const welcomePage = (props: WelcomeProps): void => {
+  const container = document.getElementById("pageContent") as HTMLElement;
+  render(pageContent(props), container);
+};
+
+export const welcome = (connection: Connection): Promise<LoginData> => {
+  const retryOnError = async (result: LoginFlowResult): Promise<LoginData> => {
```

Comment:
this is almost the same piece of code as above with `authenticateBox`, maybe it can be refactored (later on) to avoid code duplication?

Reply:
Yes, definitely; want to pull the error handling out in both and also avoid nested function calls on error (right now we're recursively calling the function which isn't nice)